### PR TITLE
Separate MultiIndex names from levels

### PR DESCRIPTION
**Pull request by topper-123**
- [x] progress towards #27138 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

In #27138 I proposed doing some changes to ``MultiIndex``, so that the index type can have its data collected in ``_data`` as type ``List[Categorical]``,+ adding ``MultiIndex.arrays`` in order to access each full level as zero-copy ``Categorical``.

This is the first part of that proposal, and drops setting the names on the  ``levels[x].name`` attribute and instead sets the names on the ``MultiIndex._names`` attribute.

This PR is a minorly backward-breaking change (so would be good to get into 0.25), while the followup will not break anything.

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -259,6 +259,7 @@ def __new__(
         result._set_levels(levels, copy=copy, validate=False)
         result._set_codes(codes, copy=copy, validate=False)
 
+        result._names = [None for _ in levels]
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: ``[None] * len(levels)``

File: pandas/core/reshape/reshape.py [hunk 2]
```diff
@@ -260,10 +260,13 @@ def get_new_values(self):
     def get_new_columns(self):
         if self.value_columns is None:
             if self.lift == 0:
-                return self.removed_level
+                lev = self.removed_level._shallow_copy()
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: why wouldn't you do
```lev = self.removed_level._shallow_copy(name=self.removed_name)``` ?

File: pandas/core/reshape/reshape.py [hunk 3]
```diff
@@ -658,7 +663,9 @@ def _convert_level_number(level_num, columns):
         new_names = this.columns.names[:-1]
         new_columns = MultiIndex.from_tuples(unique_groups, names=new_names)
     else:
-        new_columns = unique_groups = this.columns.levels[0]
+        new_columns = this.columns.levels[0]._shallow_copy()
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: use name= here

File: pandas/core/reshape/reshape.py [hunk 4]
```diff
@@ -302,7 +305,9 @@ def get_new_index(self):
             lev, lab = self.new_index_levels[0], result_codes[0]
             if (lab == -1).any():
                 lev = lev.insert(len(lev), lev._na_value)
-            return lev.take(lab)
+            new_index = lev.take(lab)
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: I would use .rename()

File: pandas/core/reshape/reshape.py [hunk 5]
```diff
@@ -260,10 +260,13 @@ def get_new_values(self):
...
         if self.value_columns is None:
             if self.lift == 0:
-                return self.removed_level
+                lev = self.removed_level._shallow_copy()
+                lev.name = self.removed_name
+                return lev
 
-            lev = self.removed_level
-            return lev.insert(0, lev._na_value)
+            lev = self.removed_level.insert(0, item=self.removed_level._na_value)
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: I would use .rename()

File: pandas/tests/frame/test_alter_axes.py [hunk 6]
```diff
@@ -979,7 +979,7 @@ def test_reset_index(self, float_frame):
         ):
             values = lev.take(level_codes)
             name = names[i]
-            tm.assert_index_equal(values, Index(deleveled[name]))
+            tm.assert_index_equal(values, Index(deleveled[name]), check_names=False)
```
Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: why is this changed?

**Review by jreback — CHANGES_REQUESTED** [review]
in MultiIndex, no problem setting the level.name attribute, but if you are in frame.py or reshape.py I would avoid doing this and instead use the name= parameter in ._shallow_copy() or use .rename()? (on a level)

File: pandas/core/reshape/reshape.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/reshape/reshape.py]: _shallow_copy and rename and other indurect methods to set the .name all call ._set_names, which does a lot of checks. Those checks are not needed in these cases, as the name has already been validated.

Perhaps have a fastpath parameter in _set_names?

File: pandas/tests/frame/test_alter_axes.py (hunk 6 above)
Comment (topper-123) [reply pandas/tests/frame/test_alter_axes.py]: ``lev.take(level_codes)`` doesn't provide a name any more, while a rest index does provides its Series with a name (as it should.

I've added a test ``assert values.name is None`` to make this more explicit.

**Comment by topper-123** [comment]
I’ve changed how the name is set. This is a bit slower (many checks that are not needed), but than could be fixed seperately in a later PR.

File: pandas/tests/test_multilevel.py [hunk 7]
```diff
@@ -1609,12 +1607,12 @@ def test_constructor_with_tz(self):
         )
 
         result = MultiIndex.from_arrays([index, columns])
-        tm.assert_index_equal(result.levels[0], index)
-        tm.assert_index_equal(result.levels[1], columns)
+        tm.assert_index_equal(result.levels[0], index, check_names=False)
```
Comment (jreback) [inline pandas/tests/test_multilevel.py]: I now find these tests very confusing that we lose the names on the levels themselves. (I know that's the point of this PR).

Comment (jreback) [reply pandas/tests/test_multilevel.py]: maybe add a .set_names(index.name) (for example) and remove the check_names arg (so its the default of True)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/test_multilevel.py (hunk 7 above)
Comment (topper-123) [reply pandas/tests/test_multilevel.py]: I've made changes so we avoid ``check_names`` is not used anymore (so is implicitly ``True``.

**Comment by topper-123** [comment]
Is this ok? I'd like to get this in 0.25, as this is a breaking change. I'll add a 0.25 label as a reminder.

The rest of #27138 will be non-breaking, so can go in later, if needed.

**Comment by TomAugspurger** [comment]
I haven’t looked, but we shouldn’t merge breaking changes in the release candidate.


________________________________

**Comment by topper-123** [comment]
Yeah, I understand that, but the followups to this PR will in addition to the benefits mentioned in #27138 also allow some nice simplifications of MultiIndex (by delegating all single-level checks to Categorical) and I assume release of 0.25 will mean a stop to breaking changes for a while,, because next up will be 1.0?

**Comment by jreback** [comment]
@TomAugspurger I don't believe we have held off on merging even breaking changes to an RC. I don't see this as a big deal and would merge as is.

**Comment by TomAugspurger** [comment]
I haven't had a chance to look (and won't this week), but if we're merging
API changes in RC0 then we'll need a second RC.

**Comment by topper-123** [comment]
Ping.

File: doc/source/whatsnew/v0.25.0.rst [hunk 8]
```diff
@@ -250,6 +250,18 @@ is respected in indexing. (:issue:`24076`, :issue:`16785`)
     df['2019-01-01 12:00:00+04:00':'2019-01-01 13:00:00+04:00']
 
 
+.. _whatsnew_0250.api_breaking.MultiIndex._names:
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you move to 1.0

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by topper-123** [comment]
I've moved the whatsnew section to 1.0.

File: doc/source/whatsnew/v1.0.0.rst [hunk 9]
```diff
@@ -31,7 +31,16 @@ Other enhancements
...
 .. _whatsnew_1000.api_breaking:
 
--
+
+.. _whatsnew_1000.api_breaking.MultiIndex._names:
+
+``MultiIndex.levels`` do not hold level names any longer
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+- A :class:`MultiIndex` previously stored the level names as attributes of each of its
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: can you add the issue number here (or this PR number)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@topper-123 lgtm (tiny comment). @TomAugspurger any objections?

**Comment by TomAugspurger** [comment]
Question on the broad goal of https://github.com/pandas-dev/pandas/issues/27138: IIUC, the motivation is for MultiIndex to be backed by a `_data: List[Categorical]`. I don't necessarily agree that that goal requires API breaking changes here.

Why can `.levels` not return something like

```python
@property
def levels(self):
    return FrozenList(pd.Index(self._data[i], name=self.names[i] for i in idx.nlevels))
```

**Comment by topper-123** [comment]
Your example is not equivalent to ``levels``, as ``._data[i]`` will be a ``Categorical``. Rather each level should be equivalent to ``._data[i].categories``. It does not make sense that categories have a ``.name``. So it will be better in terms of having clear code paths, if we avoid having a ``.name`` in the levels.

Your example would probably work, but we'd always have to be mindful that levels are different than categories, and we shoul never use levels internally (because they'd be more expensive to construct than accessing the categories). And for that extra complexity we would't get any gain in return, so I think it's better to just seperate names and levels.

**Comment by TomAugspurger** [comment]
> Your example is not equivalent to levels, as ._data[i] will be a Categorical. Rather each level should be equivalent to ._data[i].categories.

You're right, my apologies. I think the broader point about `.levels` wrapping the source of truth `._data[I]` in an Index with a name stands.

> but we'd always have to be mindful that levels are different than categories, and we shoul never use levels internally

Isn't the plan to use `MutliIndex._data` internally rather than `MultiIndex.levels`?

> And for that extra complexity we would't get any gain in return, so I think it's better to just seperate names and levels.

But we don't break API, which IMO is worth it when the change *seems* unnecessary.

**Comment by topper-123** [comment]
> Isn't the plan to use MutliIndex._data internally rather than MultiIndex.levels?

Yes, that's the plan. My point was that if ``name`` can be accessed through the levels, we'd have to be careful not to use levels in the future, because it's a list of indexes and indexes are expensive to construct compared to a pure access to ``_data[i].categories``.

Also, when constructing MultiIndex instances we'd have three places to check the names (``name``, ``names`` and in ``levels``) compared to two if we drop having names in levels.

> But we don't break API, which IMO is worth it when the change seems unnecessary.

That's a fair point, though I think this is a worthwhile simplification of the API.

This perhaps leads to a question if the API is the right one and if multiIndex should accept categories instead of levels, so concepts known from other places of pandas could be reused in MultiIndex.

**Comment by TomAugspurger** [comment]
> Also, when constructing MultiIndex instances we'd have three places to check the names (name, names and in levels) compared to two if we drop having names in levels.

That's the status quo right? If we want to deprecate passing names in one of those places then by all means let's have that conversation.

> This perhaps leads to a question if the API is the right one and if multiIndex should accept categories instead of levels, so concepts known from other places of pandas could be reused in MultiIndex.

Perhaps, I don't have an opinion at this point.

**Comment by topper-123** [comment]
This is a simplification of the code base, as it gives better seperation of concern.

Also, MultiIndex gets more similar to other index types (and Series etc.), where the name is NOT part of the underlying data structure, as it currently is for MultiIndex. E.g. we in general have ``idx._data`` storing the data and ``idx.name`` (or ``._name``) storing the name and the name is (correctly) not stored in the data.

**Comment by TomAugspurger** [comment]
> This is a simplification of the code base, as it gives better seperation of concern.

I don't disagree, but is it an API change?

**Comment by topper-123** [comment]
> I don't disagree, but is it an API change?

Do you mean in the whatsnew? I would say it's an API change, as previously you could access names by ``midx.levels[1].name``, which no longer works.

OTOH perhaps it's only me who has been aware of this and others have seen it as an implementation detail? It does not seem to be documented for example [in the docs](https://dev.pandas.io/user_guide/advanced.html#hierarchical-indexing-multiindex)

**Comment by TomAugspurger** [comment]
> OTOH perhaps it's only me who has been practically aware of this and others have seen it as an implementation detail?

No, it's a public attribute. I don't think we can call that an implementation detail. I know I've used it in the past.

**Comment by topper-123** [comment]
So, do you agree with this code change overall and pull it in, @TomAugspurger?

**Comment by TomAugspurger** [comment]
I'm not too familiar with MultiIndex's internals, but it seems like a reasonable refactor. We just need to deprecate acceding `.levels[x].name` with something like https://github.com/pandas-dev/pandas/pull/27242#issuecomment-516076989

**Comment by topper-123** [comment]
Oh, doing that will likely cause large slowdowns, as we will need to create new indexes for each new ``levels`` access, and index instantiation is very slow.

**Comment by TomAugspurger** [comment]
Can you measure the slowdown?

Presumably that can be cached to offset some of the slowdown, if it's actually a problem.

**Comment by topper-123** [comment]
```python
>>>  x = ['a', 'b', 'c', 'd']
>>> %timeit pd.Index(x)
91.7 µs ± 2.72 µs
>>> %timeit pd.CategoricalIndex(x)
507 µs ± 9.98 µs
>>> %timeit pd.date_range(2010, periods=4)
238 µs ± 4.53 µs
```

So for a 4-level MultiIndex, probably costs a millisecond each time ``levels`` would be accessed which seems too much, given that this is index operations.
 
Architecturally, we'd also have to consider that ``midx.levels[2].name = 'abc'`` is currently possible, but would not work under your proposal (without adding complexity).

**Comment by TomAugspurger** [comment]
Is that a representative benchmark? We're not passing a list to the index constructor, we already have a valid index, we're just adding a name.

```python
In [53]: idx = pd.MultiIndex.from_product([['a', 'b', 'c'], [1, 2, 3]], names=['a', 'b'])

In [54]: %timeit idx.levels[0]
626 ns ± 19.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [55]: %timeit idx.levels[0]._shallow_copy(name='a')
4.98 µs ± 101 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

For user-facing code, I think that's acceptable. If it proves to be a slowdown internally, we can have a private `._levels` attribute that doesn't do the wrapping with the name. But I suspect it's not going to be a problem in practice.

> Architecturally, we'd also have to consider that midx.levels[2].name = 'abc' is currently possible, but would not work under your proposal (without adding complexity).

Is the additional complexity reflecting that update in MultiIndex.names? That doesn't seem too difficult, and I'm quite happy to deprecate this behavior in favor of updating names.

**Comment by topper-123** [comment]
Ok, hadn't considered ``_shallow_copy``, though there will still be an extra step for instantiating the ``._engine``, which is O(N) for the size of the level, so will be noticable for large levels, but not small ones:

```python
>>> x = pd.CategoricalIndex(np.arange(1_000_000))
>>> %timeit x.get_loc(999_999)
21.7 µs ± 319 ns per loop
>>> %timeit x._shallow_copy().get_loc(999_999)
2.61 ms ± 31.7 µs per loop
```

> Is the additional complexity reflecting that update in MultiIndex.names? That doesn't seem too difficult, and I'm quite happy to deprecate this behavior in favor of updating names.

Not sure I understand you here. Do you mean drop the behaviour?

**Comment by TomAugspurger** [comment]
I mean we can deprecate setting of the names via the levels.

**Comment by topper-123** [comment]
But a level is just a index, so we can't deprecate setting the name on a level without doing the same for indexes. Are we miscommunicating?

**Comment by TomAugspurger** [comment]
No, I was likely just wrong about being able to deprecate setting `.levels[i].name` and have it be reflected in `MultiIndex.names`.

> Architecturally, we'd also have to consider that midx.levels[2].name = 'abc' is currently possible, but would not work under your proposal (without adding complexity).

Do you have a plan for how to achieve that? I'm curious how complex it is.

**Comment by topper-123** [comment]
> Do you have a plan for how to achieve that? I'm curious how complex it is.

Not really. It doesn't make sense to change how the index name is set, so probably not practically possible. This change will make the names more similar to other names, so overall this will be a nice gain, IMO.

**Comment by TomAugspurger** [comment]
Are we comfortable with that API change (@jreback, @WillAyd, @jorisvandenbossche)?

```python
In [46]: idx = pd.MultiIndex.from_product([['a', 'b'], [1, 2, 3]], names=['A', 'B'])

In [47]: idx
Out[47]:
MultiIndex([('a', 1),
            ('a', 2),
            ('a', 3),
            ('b', 1),
            ('b', 2),
            ('b', 3)],
           names=['A', 'B'])

In [48]: idx.levels[0].name = 'C'  # this will silently fail to update idx.names
```

I'm not sure yet how I feel about it.

**Comment by topper-123** [comment]
This is the same as for ``CategoricalIndex`` already:

```python
>>> ci = pd.CategoricalIndex(list('abc'), name='x')
>>> ci.categories.name = 'y'
>>> ci
CategoricalIndex(['a', 'b', 'c'], categories=['a', 'b', 'c'], ordered=False, name='x', dtype='category')
>>> ci.categories.name
'y'
```

In general, for all pandas container types, the name attribute is never set on inner data, except for ``MultiIndex``. This PR will just make things better for ``MultiIndex``, giving it clearer seperation of concern also (IMO, of course:-))

**Comment by WillAyd** [comment]
I'm assuming users will still be able to do something like:

```python
idx.names[0] = "C"
```

Instead of the more verbose

```python
idx.levels[0].name = "C"
```

that we are talking about right? Or is that also not possible here?

**Comment by topper-123** [comment]
Yes, that will be possible (and the intended way to do that).

That current ``idx.names[0] = "C"`` being the same as ``idx.levels[0].name = "C"`` is probably not very intuitive for the user, IMO.

**Comment by TomAugspurger** [comment]
> That current idx.names[0] = "C" being the same as idx.levels[0].name = "C" is probably not very intuitive for the user, IMO.

That's not possible today, right? Since `id.names` is a FrozenList.

**Comment by WillAyd** [comment]
Do we have a test that hits that? I tried this locally to no avail:

```python
>>> mi = pd.MultiIndex.from_product((("a", "b"), ("c", "d")), names=["foo", "bar"])
>>> df = pd.DataFrame(np.ones((4,4)), index=mi)
>>> df.index.names[1] = "qux"
----> 1 df.index.names[1] = "qux"

~/clones/pandas/pandas/core/indexes/frozen.py in _disabled(self, *args, **kwargs)
    108         """This method will not function because object is immutable."""
    109         raise TypeError(
--> 110             "'%s' does not support mutable operations." % self.__class__.__name__
    111         )
    112

TypeError: 'FrozenList' does not support mutable operations.
```

I guess thinking through it more I'm somewhat concerned about how we manage mutability here. The main thing I'm not sure of is if doing something like:

```python
>>> df.index.names[1] = "qux"
```

Would then get reflected if a user did

```python
>>> df.reset_index(level=1)
```

When that particular level gets moved to the columns of the DataFrame, or if the original name is what gets carried through there

**Comment by topper-123** [comment]
Looking again, that actually doesn't work currently in the PR, becauses ``.names`` is a ``FrozenList``. So currently only ``idx.names = new_names`` works... I'll see if I can work around that. Personally I would't mind too much if setting single names isn't possible as it'd be similar to how you can change all thr columns (``df.columns = new_columns``), but you can't change a individual label in a column (``df.columns[0] = 'a'  # does not work``).

**Comment by TomAugspurger** [comment]
So, summarizing our options

1. This PR as is: Breaking API to `MultiIndex.levels[n].name`. Getting a name that way will always be `None`. Setting a name there will silently not be reflected in `MultiIndex.names`
2. Restore `MultiIndex.levels[n].name` returning `MultiIndex.names[n]`. Setting `MultiIndex.levels[n].name` would still silently fail to be reflected in `MultiIndex.names`.
3. No changes from master. Does this prohibit the `List[Categorical]` refactor?

I'm not comfortable with option 1. I think it's too large of a breaking change.

I think that number 2 is the best option. It's still an API change on setting the names, but I think it's small enough to be OK. It's just unfortunate that the setting of the name would be silently not reflected in `MultiIndex.names`, and this introduces an asymmetry between getting and setting.

**Comment by WillAyd** [comment]
I'm not sure option 2 is worth doing alone though think it would be great to decouple the managing of names from the MultiIndex / Index . Is there a simple way to detect whether or not an Index is part of a MultiIndex and if so throw a warning about setting the name? 

Maybe we need some kind of `Index.set_name(names, level=0)` function to manage this appropriately (also would help #19082). Don't want to get too sidetracked but there are some definite corner cases here

**Comment by TomAugspurger** [comment]
I think `MultiIndex.rename(name, level=0)` already works for the specific
issue of updating a single name.

```
In [2]: idx = pd.MultiIndex.from_product([['a', 'b'], [1, 2, 3]],
names=['A', 'B'])

In [3]: idx.rename('C', level=0).names
Out[3]: FrozenList(['C', 'B'])
```

**Comment by jreback** [comment]
why couldn’t option 2 show deprecation warnings on both getting and setting? these are just property accessors on a subclass of FrozenList

**Comment by TomAugspurger** [comment]
The issues with the Indexes inside the FrozenList

```python
In [7]: idx = pd.MultiIndex.from_product([['a', 'b'], [1, 2, 3]], names=['A', 'B'])

In [8]: idx.levels[0]
Out[8]: Index(['a', 'b'], dtype='object', name='A')
```

So we would want to deprecate setting / getting the `.name` on `Out[8]`.

**Comment by jreback** [comment]
did we have consensus on this? @topper-123 can you summarize the arguments / issues here.

**Comment by topper-123** [comment]
My summary:

On all other index types except MultiIndex, the name is set on the index and not in its content. Setting the names on the MultiIndex instead of in the ``.labels`` will simplify a lot of things about MultiIndexes and makes MultiIndex more similar to the rest of pandas, especially ``CategoricalIndex``, which in many ways can be seen as a single level MultiIndex.

For example, one major architectural difference between a ``Categorical.categories`` and a level in a MultiIndex is that the name of the MultiIndex level is stored in the level i.e. in ``MultiIndex.levels[x]``, but not in the ``Categorical.categories`` . This makes storing the MultiIndex codes and levels as a Categorical internally much more tricky than needed, as the programmer has to be mindful of where the name is located.

I'm convinced this is a needed improvement to the architechture of the MultiIndex and should be done for v1.0, and it will make working with MultiIndex much easier in the future.

**Comment by TomAugspurger** [comment]
@topper-123 is my option 2 in https://github.com/pandas-dev/pandas/pull/27242#issuecomment-527543183 viable?

I think we're agreed that

1. This is a valuable refactor
2. Breaking *setting* MultiIndex.levels[n].name is OK

The point of disagreement is whether this refactor is worth breaking getting `MultiIndex.levels[n].name`, or whether we can deprecate getting the names from there rather than `MultiIndex.names[i]`.

**Comment by topper-123** [comment]
Alright, that's great.

Regarding getting the level name from ``MultiIndex.levels[n]``, as you know that ``MultiIndex.levels`` is a list og ``Index`` (or subclass of ``Index``) instances, so we couldn't change the levels without changing ``Index`` also, which probably isn't a  good idea.

An idea though (changing Index to dissallow setting the name):

```python
class Index(...)
    _set_name_allowed = True
    _name = None

    @property
    def name(self):
        return self._name
    @name.setter
    def name(self, value):
        if not self._set_name_allowed:
            raise TypeError("Not allowed setting names on MultiIndex levels")
        self._set_name(...)
```

Then upon instantiating ``MultiIndex``, set each level's ``_set_name_allowed`` to False.

That's a fair amount of abstraction though, and I'm worried that it will be leaky, and this "solution" will just create different problems...

**Comment by topper-123** [comment]
@TomAugspurger, Any comment? I'm leaningon that this will just create a different set of issues, so it will be cleaner to just make it a clean cut.

File: doc/source/whatsnew/v1.0.0.rst [hunk 10]
```diff
@@ -37,6 +37,16 @@ Other enhancements
...
+  :attr:`MultiIndex.levels`. From Pandas 1.0, the names are only accessed through
+  :attr:`MultiIndex.names` (which was also possible previously). This is done in order to
+  make :attr:`MultiIndex.levels` more similar to :attr:`CategoricalIndex.categories` (:issue:`27242`:).
+
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: can u show the previous way in a code block and the new way in an ipython block

**Comment by jreback** [comment]
i am +1 on this PR as is
i don’t think we should go to crazy lengths to depreciate here
the point of a breaking change is that it is an actual break in 1.0

**Comment by TomAugspurger** [comment]
> Alright, that's great

What’s that referring to?

Has https://github.com/pandas-dev/pandas/pull/27242#issuecomment-516076989 been implemented? AFAICT, that restores getting the name from .levels[n]. Using shallow copy as in https://github.com/pandas-dev/pandas/pull/27242#issuecomment-526789771

**Comment by topper-123** [comment]
@TomAugspurger : I read your comment like you were postive to merging this, but you wanted to clarify if some trade-offs could be avoided. It was in that light I commented, and offered a possible solution (which I'm not so sure is good one).

**Comment by TomAugspurger** [comment]
OK, sorry I wasn't clear. That was meant to summarize the state of the issue. I believe that

> [Tom] The point of disagreement is whether this refactor is worth breaking getting MultiIndex.levels[n].name, or whether we can deprecate getting the names from there rather than MultiIndex.names[i].

is still the outstanding thing to resolve. I'm still against breaking `MultiIndex.names[i]`, when it seems like we have a relatively easy way to not break *getting* the names with something like https://github.com/pandas-dev/pandas/pull/27242#issuecomment-516076989.

**Comment by topper-123** [comment]
So, wrt. your proposal:

```python
@property
def levels(self):
    return FrozenList(pd.Index(self._data[i], name=self.names[i] for i in idx.nlevels))
```

As I see it, this will for practical purposes keep the names in two locations. I think that will be confusing for users, especially if it isn't followed by a deprecation message (though we could mitigate that by setting a flag, like I discussed above).

**Comment by TomAugspurger** [comment]
> As I see it, this will for practical purposes keep the names in two locations

No, they'll be stored in one place and accessible from two locations. This is the same as on master, only the source of truth and the referring location are swapped.

---

w.r.t to potentially deprecating accessing `.levels[n].name`, I think something a bit simpler than https://github.com/pandas-dev/pandas/pull/27242#issuecomment-539607526 is possible. Essentially, make `.name` of the returned object a property that warns before returning the original.

I think deprecating accessing `levels[n].name` can be done in a followup PR if that's what we want to do.

**Comment by TomAugspurger** [comment]
@topper-123 do you have thoughts on https://github.com/pandas-dev/pandas/pull/27242#issuecomment-541916305?

**Comment by topper-123** [comment]
>> As I see it, this will for practical purposes keep the names in two locations

>No, they'll be stored in one place and accessible from two locations. This is the same as on master, only the source of truth and the referring location are swapped.

But if someone does

```python
>>> mi = pd.MultiIndex.from_product([[1, 2], ['a', 'b']], names=['x', 'y'])
>>> lev = mi.levels[0]
>>> mi.set_names('z', level=0)
# then
>>> mi.names[0], lev.name
'z', 'x'
```

So the names will be stored in tow places (or users should not store individual levels seperately, which they can't be expected to know. So for this reason I think it's the most most practical to make a clean cut.

**Comment by TomAugspurger** [comment]
Ahh, a new name for `levels` is a good idea for getting around this problem.

**Comment by jreback** [comment]
+1 on @topper-123 new idea.

**Comment by TomAugspurger** [comment]
On the new name, is `.categories` what we want?

I suspect that in the near-term we'll have a `DictEncodedArray` that's like a Categorical, except it won't have the same semantics around "unobserved" categories and new categories can be added implicitly. I'm not sure what we would call the set of unique values for that thing (perhaps `categories` too).

**Comment by TomAugspurger** [comment]
I think `.categories` is fine for now.

It's a bit unfortunate that it's a `List[Categorical]` rather than an Index
like on CategoricalIndex, but that's probably OK.

**Comment by TomAugspurger** [comment]
I made https://github.com/pandas-dev/pandas/issues/29032 so that this isn't
dropped.

**Comment by jorisvandenbossche** [comment]
This is breaking pyarrow (https://issues.apache.org/jira/browse/ARROW-6922).

The above is a very long discussion (which I didn't really follow before, sorry for that), but what I somewhat understand is that for 1.0 we want to restore the `.levels` behaviour of being able to *get* the name? But why was this PR then merged? 
Or why was the suggestion of Tom to wrap the categories in an Index with the name as the `levels` (https://github.com/pandas-dev/pandas/pull/27242#issuecomment-516076989) not implemented?

**Comment by jorisvandenbossche** [comment]
If the idea is to deprecate the `.levels` attribute of a MultiIndex, I think that deserves first some separate focused discussion (as that is a big change). 

Short-term, can we add back the names to `.level`? From the discussion above, it seems this does not need to conflict with the refactor of the MultiIndex internals (for *getting* the name)

**Comment by TomAugspurger** [comment]
> The above is a very long discussion (which I didn't really follow before, sorry for that), but what I somewhat understand is that for 1.0 we want to restore the .levels behaviour of being able to get the name? But why was this PR then merged?

Right. Restoring getting is relatively straightforward. I can put a PR up with that later today.
